File: kernel/drivers/src/chrdev.zig
const c = @cImport({
    @cInclude("ferrite/errno.h");
    @cInclude("ferrite/types.h");
    @cInclude("drivers/chrdev.h");
    @cInclude("sys/file/file.h");
    @cInclude("drivers/printk.h");
});

const MAX_CHRDEV = 32;
var chrdev_table: [MAX_CHRDEV]?*const c.struct_file_operations = [_]?*const c.struct_file_operations{null} ** MAX_CHRDEV;

pub export fn register_chrdev(major: c_uint, ops: *const c.struct_file_operations) c_int {
    if (major >= MAX_CHRDEV) {
        return -c.EINVAL;
    }

    if (chrdev_table[major] != null) {
        return -c.EBUSY;
    }

    chrdev_table[major] = ops;
    return 0;
}

pub export fn get_chrdev(major: c_uint) ?*const c.struct_file_operations {
    if (major >= MAX_CHRDEV) {
        return null;
    }

    return chrdev_table[major];
}

pub export fn unregister_chrdev(major: c_uint) c_int {
    if (major >= MAX_CHRDEV) {
        return -c.EINVAL;
    }

    chrdev_table[major] = null;
    return 0;
}

File: kernel/drivers/src/chrdev.c
#include "sys/file/file.h"

#include <drivers/chrdev.h>
#include <uapi/errno.h>
#include <types.h>

#define MAX_CHRDEV 32

const struct file_operations* chrdev_table[MAX_CHRDEV] = { 0 };

int register_chrdev(unsigned int major, const struct file_operations* ops)
{
    if (major >= MAX_CHRDEV) {
        return -EINVAL;
    }

    if (chrdev_table[major] != NULL) {
        return -EBUSY;
    }

    chrdev_table[major] = ops;
    return 0;
}

struct file_operations const* get_chrdev(unsigned int major)
{
    if (major >= MAX_CHRDEV) {
        return NULL;
    }

    return chrdev_table[major];
}

int unregister_chrdev(unsigned int major)
{
    if (major >= MAX_CHRDEV) {
        return -EINVAL;
    }

    chrdev_table[major] = NULL;
    return 0;
}
